SwiftCompilerSources/Sources/Optimizer/Utilities/Test.swift

```diff
@@ -130,27 +131,61 @@ extension BridgedTestArguments {
   public var native: TestArguments { TestArguments(bridged: self) }
 }
 
-public struct TestContext {
-  public var bridged: BridgedTestContext
-  fileprivate init(bridged: BridgedTestContext) {
-    self.bridged = bridged
-  }
+/// Registration of each test in the SIL module.
+public func registerOptimizerTests() {
+  // Register each test.
+  registerFunctionTests(
+      parseTestSpecificationTest,
+      forwardingUseDefTest,
+      forwardingDefUseTest
+    )
+
+  // Finally register the thunk they all call through.
+  registerFunctionTestThunk(functionTestThunk)
 }
 
-extension BridgedTestContext {
-  public var native: TestContext { TestContext(bridged: self) }
+private func registerFunctionTests(_ tests: FunctionTest...) {
+  tests.forEach { registerFunctionTest($0) }
 }
 
-private func registerFunctionTest(
-  _ test: FunctionTest,
-  implementation: BridgedFunctionTestThunk) {
+private func registerFunctionTest(_ test: FunctionTest) {
   test.name._withStringRef { ref in
-    registerFunctionTest(ref, implementation)
+    registerFunctionTest(ref, eraseInvocation(test.invocation))
   }
 }
 
-public func registerSILTests() {
-  registerFunctionTest(parseTestSpecificationTest, implementation: { parseTestSpecificationTest.run($0, $1, $2) })
+/// The function called by the swift::test::FunctionTest which invokes the
+/// actual test function.
+///
+/// This function is necessary because tests need to be written in terms of
+/// native Swift types (Function, TestArguments, BridgedPassContext)
+/// rather than their bridged variants, but such a function isn't representable
+/// in C++.  This thunk unwraps the bridged types and invokes the real
+/// function.
+private func functionTestThunk(
+  _ erasedInvocation: UnsafeMutableRawPointer,
+  _ function: BridgedFunction, 
+  _ arguments: BridgedTestArguments, 
+  _ passInvocation: BridgedSwiftPassInvocation) {
+  let invocation = uneraseInvocation(erasedInvocation)
+  let context = FunctionPassContext(_bridged: BridgedPassContext(invocation: passInvocation.invocation))
+  invocation(function.function, arguments.native, context)
+}
+
+/// Bitcast a thin test closure to void *.
+///
+/// Needed so that the closure can be represented in C++ for storage in the test
+/// registry.
+private func eraseInvocation(_ invocation: FunctionTestInvocation) -> UnsafeMutableRawPointer {
+  return unsafeBitCast(invocation, to: UnsafeMutableRawPointer.self)
+}
+
+/// Bitcast a void * to a thin test closure.
+///
+/// Needed so that the closure stored in the C++ test registry can be invoked
+/// via the functionTestThunk.
+private func uneraseInvocation(_ erasedInvocation: UnsafeMutableRawPointer) -> FunctionTestInvocation {
```

Comment:
-> `createInvocation(fromOpaquePointer: UnsafeMutableRawPointer)`

Reply:
Let's use consistent terminology for both: `castToX(fromY:)`  https://github.com/apple/swift/pull/69069